warehouse/warehouse.go

```diff
@@ -1066,6 +1079,232 @@ func processHandler(w http.ResponseWriter, r *http.Request) {
+		PendingEvents: pendingEvents,
+	}
+
+	resBody, err := json.Marshal(res)
+	if err != nil {
+		err := fmt.Errorf("Failed to marshall pending events response : %v", err)
+		pkgLogger.Errorf("[WH]: %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.Write(resBody)
+}
+
+func getPendingStagingFileCount(sourceID string) (fileCount int64, err error) {
+	var lastStagingFileID int64
+	sqlStatement := fmt.Sprintf(`SELECT end_staging_file_id FROM %[1]s WHERE %[1]s.source_id='%[2]s' ORDER BY %[1]s.id DESC`, warehouseutils.WarehouseUploadsTable, sourceID)
+
+	err = dbHandle.QueryRow(sqlStatement).Scan(&lastStagingFileID)
+	if err != nil && err != sql.ErrNoRows {
+		err = fmt.Errorf("Query: %s failed with Error : %w", sqlStatement, err)
+		return
+	}
+
+	sqlStatement = fmt.Sprintf(`SELECT COUNT(*)
+                                FROM %[1]s
+								WHERE %[1]s.id > %[2]v AND %[1]s.source_id='%[3]s'`,
+		warehouseutils.WarehouseStagingFilesTable, lastStagingFileID, sourceID)
+
+	err = dbHandle.QueryRow(sqlStatement).Scan(&fileCount)
+	if err != nil && err != sql.ErrNoRows {
+		err = fmt.Errorf("Query: %s failed with Error : %w", sqlStatement, err)
+		return
+	}
+
+	return fileCount, nil
+}
+
+func getPendingUploadCount(sourceID string) (uploadCount int64, err error) {
+	sqlStatement := fmt.Sprintf(`SELECT COUNT(*)
+								FROM %[1]s
+								WHERE %[1]s.status NOT IN ('%[2]s', '%[3]s') AND %[1]s.source_id='%[4]s'
+	`, warehouseutils.WarehouseUploadsTable, ExportedData, Aborted, sourceID)
+
+	err = dbHandle.QueryRow(sqlStatement).Scan(&uploadCount)
+	if err != nil && err != sql.ErrNoRows {
+		err = fmt.Errorf("Query: %s failed with Error : %w", sqlStatement, err)
+		return
+	}
+
+	return uploadCount, nil
+}
+
+func triggerUploadHandler(w http.ResponseWriter, r *http.Request) {
+	// TODO : respond with errors in a common way
+	pkgLogger.LogRequest(r)
+
+	// read body
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		pkgLogger.Errorf("[WH]: Error reading body: %v", err)
+		http.Error(w, "can't read body", http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	// unmarshall body
+	var triggerUploadReq warehouseutils.TriggerUploadRequestT
+	err = json.Unmarshal(body, &triggerUploadReq)
+	if err != nil {
+		pkgLogger.Errorf("[WH]: Error unmarshalling body: %v", err)
+		http.Error(w, "can't unmarshall body", http.StatusBadRequest)
+		return
+	}
+
+	sourceID := triggerUploadReq.SourceID
+	destID := triggerUploadReq.DestinationID
+
+	// return error if source id and dest id is empty
+	if sourceID == "" && destID == "" {
+		err := fmt.Errorf("Empty source and destination id")
+		pkgLogger.Errorf("[WH]: trigger upload : %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	wh := make([]warehouseutils.WarehouseT, 0)
+
+	if sourceID != "" && destID == "" {
+		// get all wh destinations for given source id
+		destinationsMapLock.Lock()
+		for _, dst := range destinationsMap {
```

Comment:
this will only give last source of a given destination. some sources might be missing if we use this map.